src/Domain/Structure/SegmentId.hpp

```diff
@@ -115,45 +88,18 @@ class SegmentId {
   /// Does the segment overlap with another?
   bool overlaps(const SegmentId& other) const noexcept;
 
- private:
-  template <size_t VolumeDim>
-  friend class ElementId;
-
-  /// Construct a SegmentId that stores a block ID and a grid index in addition
-  /// to the index within a block. Storing the data here makes it easier for
-  /// ElementId to satisfy the size requirements, because it just has to store
-  /// `VolumeDim` SegmentIds. ElementId is responsible for managing the block ID
-  /// and the grid index of its SegmentIds.
-  ///
-  /// \see ElementId
-  SegmentId(size_t block_id, size_t refinement_level, size_t index,
-            size_t grid_index) noexcept;
-  size_t block_id() const noexcept { return block_id_; }
-  void set_block_id(const size_t block_id) noexcept {
-    ASSERT(block_id < two_to_the(block_id_bits),
-           "Block id out of bounds: " << block_id << "\nMaximum value is: "
-                                      << two_to_the(block_id_bits) - 1);
-    block_id_ = block_id;
-  }
-  size_t grid_index() const noexcept { return grid_index_; }
-  void set_grid_index(const size_t grid_index) noexcept {
-    ASSERT(grid_index < two_to_the(grid_index_bits),
-           "Grid index out of bounds: " << grid_index << "\nMaximum value is: "
-                                        << two_to_the(grid_index_bits) - 1);
-    grid_index_ = grid_index;
-  }
+  // NOLINTNEXTLINE
+  void pup(PUP::er& p) noexcept;
 
-  unsigned block_id_ : block_id_bits;
+ private:
+  static constexpr size_t refinement_bits = 4;
+  static constexpr size_t max_refinement_level = 16;
+  static_assert(two_to_the(refinement_bits) >= max_refinement_level,
+                "Not enough bits to represent all refinement levels");
   unsigned refinement_level_ : refinement_bits;
   unsigned index_ : max_refinement_level;
```

Comment:
I suppose these could just be `size_t`s now to make the class even simpler?

Reply:
This works but I had to fix a bug in our existing code, which I did in a new commit